File: src/order/domain/order_repository.py
"""OrderRepository interface — abstraction for persisting Order aggregates.

The domain layer defines the interface; infrastructure provides implementations.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import List, Optional

from src.order.domain.order import Order


class OrderRepository(ABC):
    """Repository contract for the Order aggregate."""

    @abstractmethod
    def next_id(self) -> str:
        """Generate the next sequential order ID (e.g. ORD-0001)."""

    @abstractmethod
    def save(self, order: Order) -> None:
        """Persist an order (insert or update)."""

    @abstractmethod
    def find_by_id(self, order_id: str) -> Optional[Order]:
        """Return an order by its ID, or None if not found."""

    @abstractmethod
    def find_all(self) -> List[Order]:
        """Return all persisted orders."""

File: src/order/domain/OrderRepository.ts
import { Order } from "./Order.js";

/**
 * OrderRepository — Repository Interface
 *
 * Abstraction for persisting and retrieving Order aggregates.
 * The domain layer defines this interface; infrastructure provides the implementation.
 * No database-specific details leak into the domain.
 */
export interface OrderRepository {
  /** Save a new order to the store. */
  save(order: Order): Promise<void>;

  /** Find an order by its ID. Returns null if not found. */
  findById(orderId: string): Promise<Order | null>;

  /** Return all orders. */
  findAll(): Promise<Order[]>;

  /** Generate the next sequential order ID (e.g. ORD-0001, ORD-0002). */
  nextId(): Promise<string>;
}
